### PR TITLE
CI Enhancement: Supporting CI/CD with GitHub Actions

### PR DESCRIPTION
**Pull request by ercl**
This PR implements the transition from CircleCI to GitHub Actions for [issue 287](https://github.com/open-telemetry/opentelemetry-ruby/issues/287). 

I implemented a new GitHub Actions workflow to run all the jobs and tests CircleCI did. There are a couple of things to note about the workflow:

1. GitHub Actions does not support reusable groups of commands yet, so I moved the commands into bash scripts to avoid copy/pasting for each job
2. GitHub Actions only supports a subset of regex, so the tag filter is not identical ([filter pattern documentation](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet))
    1. Filters `opentelemetry-*/v[0-9]*`  instead of `^opentelemetry-.+\/v\d.*$`
    2. `*` is used over `.+` for the name because `.` is not supported and square brackets only support the characters `0-9, a-z, A-Z`, which excludes names with hyphens (e.g opentelemetry-adapters-all)
    3. `opentelemetry-/v0.1.2` is valid (can have no name)
3. GitHub Actions does not support filtering on a per-job level. 
    1. Jobs run on every push and pull request
    2. Release job runs when a pushed tag begins with `opentelemetry-` and matches the tag filter.
4. GitHub Actions requires file permissions to be updated at the start of each job when using CircleCI images
5. `OPENTELEMETRY_RUBYGEMS_API_KEY` and `OPENTELEMETRY_RELEASES_ENABLED` need to be saved as Secrets within the repo settings for the release job to complete successfully.

I also changed the name of the environmental variable from `CIRCLE_TAG` to `GITHUB_ACTIONS_TAG` so the release job works properly.

### CODE REVIEW

**Comment by mwear** [comment]
@dazuma If you have a moment, would you mind taking a look at this since you know the most about the current setup?

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        run: sudo chmod -R 777 $GITHUB_WORKSPACE /github /__w/_temp
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Store tag as environmental variable
+        shell: bash
+        run: echo ::set-env name=GITHUB_ACTIONS_TAG::${GITHUB_REF/refs\/tags\//}
+      - name: rake-release
+        run: ./ci/rake_release.sh
+      - name: rake-test-appraisal
+        run: ./ci/rake_test_appraisal.sh
```
Comment (dazuma) [inline .github/workflows/ci.yml:137]: Remove this step. (The release job already depends on the test jobs, e.g. `test-ruby25` which should already include the appraisal steps.)

Also remove the env vars and services related to appraisal tests.

File: Rakefile [hunk 2]
```diff
@@ -165,7 +165,7 @@ def push_release
 end
 
 def parse_tag
-  tag = ENV['CIRCLE_TAG']
+  tag = ENV['GITHUB_ACTIONS_TAG']
```
Comment (dazuma) [inline Rakefile:168]: Can you support both (e.g. `ENV['CIRCLE_TAG'] || ENV['GITHUB_ACTIONS_TAG']`) until the circle jobs are removed?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (dazuma) [inline .github/workflows/ci.yml:137]: Also please make sure there's a trailing newline for all files.

File: ci/rake_release.sh [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
```
Comment (dazuma) [inline ci/rake_release.sh:2]: Please:

```sh
set -e
```

In all bash scripts used in CI. This ensures that when a nonzero exit code is detected, the script stops and exits with that same code. (Otherwise, we'll see test failures in the logs but the actual CI job won't fail.)

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    container: circleci/ruby:2.7-buster
+    steps:
+      - name: Setup file system permissions
+        run: sudo chmod -R 777 $GITHUB_WORKSPACE /github /__w/_temp
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: rake-test
+        run: ./ci/rake_test.sh
+  test-jruby:
+    name: JRuby-latest Test
```
Comment (dazuma) [inline .github/workflows/ci.yml:79]: This name isn't consistent with the others. Maybe `Test JRuby-latest`?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+      - name: Setup file system permissions
+        run: sudo chmod -R 777 $GITHUB_WORKSPACE /github /__w/_temp
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: rake-test
+        run: ./ci/rake_test.sh
+      - name: rake-test-appraisal
+        run: ./ci/rake_test_appraisal.sh        
+  release:
+    name: Test Release
```
Comment (dazuma) [inline .github/workflows/ci.yml:106]: Just `Release` for the name? This job doesn't actually run tests.

**Comment by dazuma** [comment]
I've also added the two needed secrets.

Currently, I have releases *disabled* in GitHub Actions. (i.e. `OPENTELEMETRY_RELEASES_ENABLED` is `false` in GitHub Actions but `true` in CircleCI). When we take the next step of removing Circle, we need to set `OPENTELEMETRY_RELEASES_ENABLED` to `true` in GitHub Actions.

File: ci/rake_release.sh (hunk 3 above)
Comment (ercl) [reply ci/rake_release.sh:2]: Done! This revealed some errors with the mysql service which I also fixed.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (ercl) [reply .github/workflows/ci.yml]: That's a good point. Fixed!

File: Rakefile (hunk 2 above)
Comment (ercl) [reply Rakefile]: Yes, I've added the CircleCI tag back in in the new commit.

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,14 +4,6 @@ All notable changes to this project will be documented in this file.
...
 ## Unreleased
 
-## Alpha v0.5.1
-
-### Bug Fix
-
-* [#299](https://github.com/open-telemetry/opentelemetry-ruby/pull/299) Fix probabilility sampler. ([@fbogsany](https://github.com/fbogsany))
-
-* [#301](https://github.com/open-telemetry/opentelemetry-ruby/pull/301) Fix require 'opentelemetry/instrumentation'. ([@fbogsany](https://github.com/fbogsany))
-
```
Comment (dazuma) [inline CHANGELOG.md:14]: Restore this changelog entry.

File: api/lib/opentelemetry/version.rb [hunk 7]
```diff
@@ -6,5 +6,5 @@
 
 module OpenTelemetry
   ## Current OpenTelemetry version
-  VERSION = '0.5.1'
+  VERSION = '0.4.0'
```
Comment (dazuma) [inline api/lib/opentelemetry/version.rb:9]: Revert this change.

File: sdk/lib/opentelemetry/sdk/trace/samplers/probability_sampler.rb [hunk 8]
```diff
@@ -16,7 +16,7 @@ class ProbabilitySampler
 
           def initialize(probability, ignore_parent:, apply_to_remote_parent:, apply_to_all_spans:)
             @probability = probability
-            @id_upper_bound = (probability * (2**64 - 1)).ceil
+            @id_upper_bound = format('%016x', (probability * (2**64 - 1)).ceil)
```
Comment (dazuma) [inline sdk/lib/opentelemetry/sdk/trace/samplers/probability_sampler.rb:19]: revert

File: sdk/lib/opentelemetry/sdk/trace/samplers/probability_sampler.rb [hunk 9]
```diff
@@ -55,7 +55,7 @@ def sample_trace_id_for_child?(parent_context, trace_id)
           end
 
           def sample_trace_id?(trace_id)
-            @probability == 1.0 || trace_id[8, 8].unpack1('Q>') < @id_upper_bound
+            @probability == 1.0 || trace_id[16, 16] < @id_upper_bound
```
Comment (dazuma) [inline sdk/lib/opentelemetry/sdk/trace/samplers/probability_sampler.rb:58]: revert

File: sdk/test/opentelemetry/sdk/trace/samplers_test.rb [hunk 10]
```diff
@@ -160,9 +160,7 @@
   end
 
   def trace_id(id)
-    first = id >> 64
-    second = id & 0xffff_ffff_ffff_ffff
-    [first, second].pack('Q>Q>')
+    format('%032x', id)
```
Comment (dazuma) [inline sdk/test/opentelemetry/sdk/trace/samplers_test.rb:163]: revert

File: exporters/jaeger/lib/opentelemetry/exporters/jaeger/version.rb [hunk 11]
```diff
@@ -8,7 +8,7 @@ module OpenTelemetry
   module Exporters
     module Jaeger
       ## Current OpenTelemetry Jaeger exporter version
-      VERSION = '0.5.0'
+      VERSION = '0.4.0'
```
Comment (dazuma) [inline exporters/jaeger/lib/opentelemetry/exporters/jaeger/version.rb:11]: Revert. And the same for all the gemspec changes below. (Bad merge?)

File: .gitignore [hunk 12]
```diff
@@ -9,6 +9,3 @@
 
 # Appraisals
 instrumentation/**/*.gemfile.lock
-
-# Vendored gems
-**/vendor/**/*
```
Comment (dazuma) [inline .gitignore:14]: Did you mean to delete this?

**Comment by ercl** [comment]
There was a bad merge yes. I merged master into my branch, but I reverted it since it added a bunch of new changes that I thought weren't directly related to this PR. Should I keep the merge or reset to the pre-merge commit?

**Comment by dazuma** [comment]
Generally it's fine to merge master. Yes, it brings in additional changes that others have done in the meantime, but those changes _are supposed to be there_. That's the point of doing a merge: we want the end result to include both your changes and everyone else's. It looks like what happened is that when you reverted the merge, you undid everyone else's changes (i.e. your PR, when merged, includes changes that undo everyone else's PRs.)

I suggest reverting that revert commit that you did. (If you're comfortable with rebasing and branch manipulation, feel free to just remove it from the history altogether, but if not, you can just add another commit to revert the changes. We can clean up the history by squashing when we merge the PR.) If you really run into trouble, it's okay to copy your changes into a fresh PR.

**Comment by ercl** [comment]
Thanks for the explanation! I've removed the revert and merged in the newest changes from master.
